src/bindings/cpp/include/key.hpp

```diff
@@ -670,7 +670,12 @@ void Key::operator++ (int) const
  */
 void Key::operator++ () const
 {
-	ckdb::keyIncRef (key);
+	if (ckdb::keyIncRef (key) == UINT16_MAX)
+	{
+		/* if activated, nullptr can't be assigned to a Key-object
```

Comment:
At first I ran the tests in Docker with the current state of the PR-branch and all tests passed.
Then I only activated (removed comment) the line 677 in [src/bindings/cpp/include/key.hpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/bindings/cpp/include/key.hpp), so that the KeyException gets thrown if the `keyIncRef (Key *)`-Function in  [src/libs/elektra/key.c](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/libs/elektra/key.c) returns `UINT16_MAX`.

After that, I did a rebuild and ran the tests again in Docker.
Now the following tests were failing:
```
The following tests FAILED:
	 1 - testtool_automergestrategy (Failed)
	 2 - testtool_backend (Failed)
	 3 - testtool_backendbuilder (Failed)
	 5 - testtool_comparison (Failed)
	 7 - testtool_mergecases (Failed)
	 8 - testtool_mergeresult (Failed)
	 9 - testtool_mergingkdb (Failed)
	10 - testtool_metamergestrategy (Failed)
	11 - testtool_newkeystrategy (Failed)
	12 - testtool_onesidestrategy (Failed)
	13 - testtool_plugindatabase (Failed)
	15 - testtool_samemountpoint (Failed)
	16 - testtool_specreader (Failed)
	17 - testtool_umount (Failed)
	20 - testcpp_contextual_basic (Subprocess aborted)
	21 - testcpp_contextual_nocontext (Failed)
	22 - testcpp_contextual_policy (Subprocess aborted)
	23 - testcpp_contextual_thread (Subprocess aborted)
	24 - testcpp_contextual_update (Subprocess aborted)
	27 - testcpp_kdb (Failed)
	28 - testcpp_key (Failed)
	29 - testcpp_keyio (Failed)
	30 - testcpp_ks (Failed)
	31 - testcpp_ksget (Failed)
	32 - testcpp_meta (Failed)
	34 - test_key.py (Failed)
	35 - test_keyset.py (Failed)
	36 - test_kdb.py (Failed)
	38 - test_tools_backendbuilder.py (Failed)
	40 - test_tools_modules.py (Failed)
	41 - test_tools_plugindatabase.py (Failed)
	43 - test_tools_specreader.py (Subprocess aborted)
	44 - test_key.lua (Failed)
	45 - test_keyset.lua (Failed)
	46 - test_kdb.lua (Failed)
	48 - testruby_key.rb (Subprocess aborted)
	49 - testruby_keyset.rb (Subprocess aborted)
	53 - testruby_tools_backendparser.rb (Subprocess aborted)
	54 - testruby_tools_modules.rb (Subprocess aborted)
	56 - testruby_tools_specreader.rb (Subprocess aborted)
	68 - testshell_markdown_base64 (Failed)
	70 - testshell_markdown_blacklist (Failed)
	72 - testshell_markdown_blockresolver (Failed)
	74 - testshell_markdown_cache (Failed)
	75 - testmod_ccode (SEGFAULT)
	76 - testshell_markdown_ccode (Failed)
	78 - testshell_markdown_conditionals (Failed)
	79 - testshell_markdown_constants (Failed)
	80 - testmod_cpptemplate (Failed)
	81 - testshell_markdown_cpptemplate (Failed)
	84 - testshell_markdown_csvstorage (Failed)
	88 - testmod_directoryvalue (SEGFAULT)
	89 - testshell_markdown_directoryvalue (Failed)
	92 - testshell_markdown_email (Failed)
	95 - testshell_markdown_file (Failed)
	98 - testshell_markdown_gopts (Failed)
	103 - testshell_markdown_hexnumber (Failed)
	105 - testshell_markdown_hosts (Failed)
	108 - testshell_markdown_ipaddr (Failed)
	111 - testmod_kconfig (Failed)
	112 - testshell_markdown_kconfig (Failed)
	115 - testshell_markdown_length (Failed)
	116 - testshell_markdown_line (Failed)
	122 - testshell_markdown_macaddr (Failed)
	123 - testshell_markdown_mathcheck (Failed)
	126 - testshell_markdown_mini (Failed)
	129 - testshell_markdown_mmapstorage (Failed)
	131 - testshell_markdown_mozprefs (Failed)
	132 - testshell_markdown_multifile (Failed)
	133 - testshell_markdown_network (Failed)
	136 - testshell_markdown_ni (Failed)
	138 - testshell_markdown_path (Failed)
	140 - testshell_markdown_process (Failed)
	142 - testshell_markdown_quickdump (Failed)
	145 - testshell_markdown_range (Failed)
	147 - testshell_markdown_reference (Failed)
	149 - testshell_markdown_rename (Failed)
	152 - testshell_markdown_rgbcolor (Failed)
	154 - testshell_markdown_shell (Failed)
	157 - testshell_markdown_specload (Failed)
	161 - testshell_markdown_template (Failed)
	163 - testshell_markdown_toml (Failed)
	165 - testshell_markdown_type (Failed)
	167 - testshell_markdown_uname (Failed)
	169 - testshell_markdown_unit (Failed)
	171 - testshell_markdown_xerces (Failed)
	172 - testmod_xerces (Failed)
	175 - testshell_markdown_yajl (Failed)
	176 - testmod_yamlcpp (Failed)
	177 - testshell_markdown_yamlcpp (Failed)
	181 - testscr_check_basic (Failed)
	182 - testscr_check_cmd_line_arguments (Failed)
	183 - testscr_check_distribution (Failed)
	185 - testscr_check_error (Failed)
	186 - testscr_check_export (Failed)
	188 - testscr_check_gen (Failed)
	189 - testscr_check_get_set (Failed)
	190 - testscr_check_import (Failed)
	191 - testscr_check_kdb_internal_check (Failed)
	192 - testscr_check_kdb_internal_suite (Failed)
	193 - testscr_check_merge (Failed)
	195 - testscr_check_mount (Failed)
	197 - testscr_check_plugins (Failed)
	199 - testscr_check_race (Failed)
	200 - testscr_check_real_world (Failed)
	201 - testscr_check_resolver (Failed)
	202 - testscr_check_spec (Failed)
	205 - testshell_db_changes (Failed)
	206 - testshell_host (Failed)
	207 - testshell_listtest (Failed)
	208 - testshell_mathcheck (Failed)
	209 - testshell_profiletest (Failed)
	210 - testshell_script (Failed)
	211 - testshell_selftest (Failed)
	212 - testshell_replay_ls (Failed)
	213 - testshell_markdown_kdb_get (Failed)
	214 - testshell_markdown_kdb_complete (Failed)
	215 - testshell_markdown_kdb_cp (Failed)
	216 - testshell_markdown_kdb_global_umount (Failed)
	217 - testshell_markdown_kdb_ls (Failed)
	218 - testshell_markdown_kdb_meta_show (Failed)
	219 - testshell_markdown_kdb_find (Failed)
	220 - testshell_markdown_tutorial_arrays (Failed)
	221 - testshell_markdown_tutorial_cascading (Failed)
	222 - testshell_markdown_tutorial_install_config (Failed)
	223 - testshell_markdown_tutorial_storage_plugins (Failed)
	224 - testshell_markdown_cmerge (Failed)
	225 - testshell_markdown_tutorial_validation (Failed)
	226 - testshell_markdown_tutorial_crypto (Failed)
	227 - testshell_markdown_msr_syntax (Failed)
	228 - testshell_markdown_readme_msr (Failed)
	229 - testshell_markdown_issue_template (Failed)
	230 - testshell_markdown_mmapstorage_shelltests (Failed)
	231 - testshell_markdown_cache_shelltests (Failed)
	269 - testkdb_conflict (Failed)
	270 - testkdb_error (Failed)
	271 - testkdb_nested (Failed)
	272 - testkdb_simple (Failed)
	273 - testkdb_contracts (Failed)
	274 - testkdb_highlevel (Failed)
```


When I tried to run the tests with the installed version, the kdb-command failed instantly with the following message:
```bash
jenkins@24559c24c9c7:~/workspace/elektra-build-docker$ kdb run_all
The command kdb run_all terminated unsuccessfully with the info:
Exception thrown by a Key, typically because you called a method on a null key. Make sure to check this with !key first
Please report the issue at https://issues.libelektra.org/
```


Of course, it would take too much time to investigate all the failing tests, but I looked at two failing tests in detail:

**29 - testcpp_keyio (file [src/bindings/cpp/tests/testcpp_keyio.cpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/bindings/cpp/tests/testcpp_keyio.cpp)):**
The following test case is failing at the line `ss << k;`:
```cpp
TEST (keyio, outmeta)
	{
		Key k ("user:/hello", KEY_META, "abc", "", KEY_META, "def", "", KEY_END);
		std::stringstream ss;
		ss.setf (std::ios_base::showbase);
		ss << k;
		EXPECT_EQ (ss.str (), "user:/hello meta:/abc meta:/def");
	}
```

If we look further into it, the implementation of the `<<`-operator in  [src/bindings/cpp/include/keyio.hpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/bindings/cpp/include/keyio.hpp) calls the function `Key::nextMeta ()` in [src/bindings/cpp/include/key.hpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/bindings/cpp/include/key.hpp) in a while loop.
```cpp
inline const Key Key::nextMeta ()
{
	const ckdb::Key * k = ckdb::keyNextMeta (key);
	return Key (const_cast<ckdb::Key *> (k));
}
```

Eventually the `keyNextMeta (Key *)` function in [src/libs/elektra/keymeta.c](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/libs/elektra/keymeta.c) gets a `NULL`-pointer from the `ksNext (KeySet *)` function in [src/libs/elektra/keyset.c](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/libs/elektra/keyset.c) and returns that `NULL`-pointer to the caller.

The following return statement with the `const_cast` now leads to the throwing of the `KeyException`, because `k` is now `NULL`.
The following constructor in [key.hpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/bindings/cpp/include/key.hpp) gets executed:
```cpp
inline Key::Key (ckdb::Key * k) : key (k)
{
	operator++ ();
}
```
Now the `NULL`-pointer gets passed to `keyIncRef (Key *)` which returns the error-value `UINT16_MAX`.


The second test-suite I looked at was 

**7 - testtool_mergecases:**
This test-suite consists of ten test cases which are defined in the file [src/libs/tools/tests/testtool_mergecases.cpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/libs/tools/tests/testtool_mergecases.cpp).
All ten test-cases failed after the line in [key.hpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/bindings/cpp/include/key.hpp) where the `KeyException` gets thrown was activated.


The problem here is the `ThreeWayMerge::detectConflicts ()` function which gets called from the `ThreeWayMerge::mergeKeySet (...)` function.
In the `detectConflicts ()`-function, the `keyMetaEqual ()`-Function gets called and the `keyMetaEqual ()`-function in [src/libs/tools/src/helper/comparison.cpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/libs/tools/src/helper/comparison.cpp) calls `Key::nextMeta ()`. There, the `keyNextMeta ()`-Functions returns `NULL` and now the same behaviour as explained for the previous tests leads to the throwing of the `KeyException`.


To conclude, the two test cases I looked at failed because the `const_cast` in `Key::nextMeta ()`.

The reason for the failing casts is the constructor in [key.hpp](https://github.com/flo91/libelektra/blob/FLOSS_T1/src/bindings/cpp/include/key.hpp) which accepts a `NULL`-pointer and executes the ++-Operator, so the ++-operation is executed on a `NULL`-pointer.